internal/controller/controllers/agent_controller.go

```diff
@@ -905,6 +915,50 @@ func (r *AgentReconciler) updateIfNeeded(ctx context.Context, log logrus.FieldLo
 	return nil
 }
 
+func (r *AgentReconciler) getIgnitionToken(ctx context.Context, log logrus.FieldLogger, agent *aiv1beta1.Agent) (string, error) {
+	ignitionEndpointTokenReference := agent.Spec.IgnitionEndpointTokenReference
+
+	if ignitionEndpointTokenReference == nil {
+		log.Info("No IgnitionEndpointTokenReference set")
+		return "", errors.New("No IgnitionEndpointTokenReference set")
+	}
+
+	secret := &corev1.Secret{}
+	secretRef := types.NamespacedName{Namespace: ignitionEndpointTokenReference.Namespace, Name: ignitionEndpointTokenReference.Name}
+	if err := r.Get(ctx, secretRef, secret); err != nil {
+		log.WithError(err).Errorf("Failed to get user-data secret %s", ignitionEndpointTokenReference.Name)
+		return "", err
+	}
+
+	ignitionConfig := &ignitionapi.Config{}
+	if err := json.Unmarshal(secret.Data["value"], ignitionConfig); err != nil {
+		log.WithError(err).Errorf("Failed to unmarshal user-data secret %s", ignitionEndpointTokenReference.Name)
+		return "", err
+	}
+
+	if len(ignitionConfig.Ignition.Config.Merge) != 1 {
+		log.Errorf("expected one ignition source in secret %s but found %d", ignitionEndpointTokenReference.Name, len(ignitionConfig.Ignition.Config.Merge))
+		return "", errors.New("did not find one ignition source as expected")
+	}
+
+	ignitionSource := ignitionConfig.Ignition.Config.Merge[0]
+
+	for _, header := range ignitionSource.HTTPHeaders {
+		if header.Name != "Authorization" {
+			continue
+		}
+		expectedPrefix := "Bearer "
+		if !strings.HasPrefix(*header.Value, expectedPrefix) {
+			log.Errorf("did not find expected prefix for bearer token in user-data secret %s", ignitionEndpointTokenReference.Name)
+			return "", errors.New("did not find expected prefix for bearer token")
+		}
+		token := (*header.Value)[len(expectedPrefix):]
+		return token, nil
+	}
+
+	return "", errors.New("Couldn't find ignition token in ignitionSource.HTTPHeaders")
```

Comment:
Consider extracting this logic to a separate function `getTokenFromHTTPHeaders`

Reply:
We plan to make the secret contain the ignitionToken directly so this logic might be deleted altogether.